File: backend/app/schemas/event.py
from pydantic import BaseModel
from typing import Optional, List, Union
from datetime import datetime

class VideoMetadataBase(BaseModel):
    video_id: str
    filename: Optional[str] = None
    frame_count: int
    fps: float
    width: int
    height: int
    duration: float
    status: Optional[str] = None

class VideoMetadataCreate(VideoMetadataBase):
    pass

class VideoMetadata(VideoMetadataBase):
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class EventBase(BaseModel):
    event_id: str
    facility_id: Optional[str] = None
    video_id: Optional[str] = None
    timestamp: float
    camera_id: Optional[str] = None
    bay_id: Optional[str] = None
    object_id: Optional[int] = None
    behaviour: str
    risk_score: float
    risk_level: str
    description: str
    reason: str
    potential_consequence: Optional[str] = None
    risk_factors_json: Optional[str] = None
    evidence_frame: Optional[str] = None
    video_reference: Optional[str] = None
    recommended_action: Optional[str] = None
    status: Optional[str] = "UNRESOLVED"
    acknowledged_by_user_id: Optional[str] = None
    acknowledged_at: Optional[datetime] = None
    model_name: Optional[str] = None
    model_version: Optional[str] = None
    inference_engine: Optional[str] = None
    confidence: Optional[float] = None
    rule_version: Optional[str] = None
    provenance_type: Optional[str] = "DEVELOPMENT_SEED"
    inference_run_id: Optional[str] = None
    frame_number: Optional[int] = None
    video_fps: Optional[float] = None
    timestamp_seconds: Optional[float] = None
    timestamp_utc: Optional[Union[datetime, str]] = None
    evidence_clip_start: Optional[float] = None
    evidence_clip_end: Optional[float] = None
    processing_latency_ms: Optional[float] = None

class EventCreate(EventBase):
    pass

class Event(EventBase):
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class InferenceRunBase(BaseModel):
    video_id: Optional[str] = None
    camera_id: Optional[str] = None
    model_name: str = "YOLO11s"
    model_version: str = "v1.4.2-tensorrt"
    inference_engine: str = "LOCAL_YOLO11"
    device: Optional[str] = "cpu"
    status: Optional[str] = "COMPLETED"
    provenance_type: Optional[str] = "REAL_INFERENCE"

class InferenceRunCreate(InferenceRunBase):
    id: str

class InferenceRun(InferenceRunBase):
    id: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/types/event.ts
export type RiskLevel = 'Low' | 'Medium' | 'High' | 'Critical';

export type IncidentStatus = 'UNRESOLVED' | 'ACKNOWLEDGED' | 'DISPATCHED';

export type ProvenanceType = 'REAL_INFERENCE' | 'DEVELOPMENT_SEED' | 'DEMO_FIXTURE' | 'PERFORMANCE_TEST' | 'UNIT_TEST';

export interface Event {
  event_id: string;
  facility_id?: string;
  video_id: string;
  timestamp: number; // in seconds
  camera_id?: string;
  bay_id?: string;
  object_id?: number;
  behaviour: string;
  risk_score: number;
  risk_level: RiskLevel;
  description: string;
  reason: string;
  tags?: string[];
  evidence_frame?: string;
  video_reference?: string;
  recommended_action?: string;
  status?: IncidentStatus | string;
  acknowledged_by_user_id?: string;
  acknowledged_at?: string;
  model_name?: string;
  model_version?: string;
  inference_engine?: string;
  confidence?: number;
  rule_version?: string;
  provenance_type?: ProvenanceType | string;
  inference_run_id?: string;
  frame_number?: number;
  video_fps?: number;
  timestamp_seconds?: number;
  timestamp_utc?: string;
  evidence_clip_start?: number;
  evidence_clip_end?: number;
  processing_latency_ms?: number;
  created_at?: string;
}

export interface InferenceRun {
  id: string;
  video_id?: string;
  camera_id?: string;
  model_name: string;
  model_version: string;
  inference_engine: string;
  device?: string;
  status?: string;
  started_at?: string;
  completed_at?: string;
  provenance_type?: ProvenanceType | string;
}

export interface EventFilterParams {
  risk_level?: RiskLevel | string;
  behaviour?: string;
  bay_id?: string;
  camera_id?: string;
  search?: string;
  limit?: number;
  skip?: number;
}

export interface VideoMetadata {
  video_id: string;
  filename?: string;
  frame_count: number;
  fps: number;
  width: number;
  height: number;
  duration: number; // in seconds
  status?: string;
  created_at?: string;
}
